core/src/main/java/org/apache/iceberg/encryption/DefaultEncryptionManager.java

```diff
@@ -0,0 +1,95 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.encryption;
+
+import java.nio.ByteBuffer;
+import java.security.SecureRandom;
+import java.util.Map;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.PropertyUtil;
+
+public class DefaultEncryptionManager implements EncryptionManager {
+  private final KeyManagementClient kmsClient;
+  private String tableKeyId;
+  private int dataKeyLength;
+  private boolean kmsGeneratedKeys;
+
+  private transient volatile SecureRandom workerRNG = null;
+
+  /**
+   * @param tableKeyId table encryption key id
+   * @param kmsClient Client of KMS used to wrap/unwrap keys in envelope encryption
+   * @param encryptionProperties encryption properties
+   */
+  public DefaultEncryptionManager(
+      String tableKeyId, KeyManagementClient kmsClient, Map<String, String> encryptionProperties) {
+    Preconditions.checkNotNull(
+        tableKeyId,
+        "Cannot create EnvelopeEncryptionManager because table encryption key ID is not specified");
+    Preconditions.checkNotNull(
+        kmsClient, "Cannot create EnvelopeEncryptionManager because kmsClient is null");
+    Preconditions.checkNotNull(
+        encryptionProperties,
+        "Cannot create EnvelopeEncryptionManager because encryptionProperties are not passed");
+    this.tableKeyId = tableKeyId;
+    this.kmsClient = kmsClient;
+    this.kmsGeneratedKeys = kmsClient.supportsKeyGeneration();
+
+    this.dataKeyLength =
+        PropertyUtil.propertyAsInt(
+            encryptionProperties,
+            EncryptionProperties.ENCRYPTION_DEK_LENGTH,
+            EncryptionProperties.ENCRYPTION_DEK_LENGTH_DEFAULT);
+  }
+
+  @Override
+  public EncryptedOutputFile encrypt(OutputFile rawOutput) {
+    if (null == workerRNG) {
+      createSecureRandomGenerator();
+    }
+
+    ByteBuffer fileDek = ByteBuffer.allocate(dataKeyLength);
+    workerRNG.nextBytes(fileDek.array());
+
+    ByteBuffer aadPrefix = ByteBuffer.allocate(EncryptionProperties.ENCRYPTION_AAD_LENGTH_DEFAULT);
+    workerRNG.nextBytes(aadPrefix.array());
+
+    KeyMetadata fileEnvelopeMetadata = new KeyMetadata(fileDek, null, aadPrefix);
+
+    return new BaseEncryptedOutputFile(rawOutput, fileEnvelopeMetadata);
+  }
+
+  @Override
+  public InputFile decrypt(EncryptedInputFile encrypted) {
+    if (encrypted.keyMetadata() == null || encrypted.keyMetadata().buffer() == null) {
+      throw new RuntimeException(
+          "Unencrypted file " + encrypted.encryptedInputFile().location() + " in encrypted table");
+    }
+
+    // Native decryption: simply return the input file. Parquet decryption will get the key from key
+    // metadata.
```

Comment:
I don't don't think this is correct.

These interfaces are intended to give you the option to encrypt/decrypt a stream. That's why they expose key metadata and an `InputFile` or `OutputFile`. You use the `InputFile` as you normally would, but it is decrypted for you as you read.

That fits with the AES GCM encryption spec. This should decrypt using AES GCM decryption. Similarly, encryption should return an encrypting stream.

For native Parquet encryption, I think the `EncryptedOutputFile` and `EncryptedInputFile` classes would need to be able to return the underlying stream as well, so that encryption can be handled by Parquet.

Reply:
Done for the  EncryptedOutputFile. For decryption, we have everything already in the Parquet reader: the underlying stream and the key metadata, so can decrypt with native Parquet.
(no need in changes in  EncryptedInputFile, and no need in calling the encryption manager).